### PR TITLE
Feiran

### PR DESCRIPTION
**Pull request by Brack-Wang**
Char Compress
Char to unsigned char
A: 00
T: 01
C: 10
G: 11

### CODE REVIEW

File: kmp_cpu.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
+#include "kmp_cpu.hpp"
+
+static void get_fail(const char *pattern, int pattern_length, int *fail) {
+    int candidate = 0;
+    fail[0] = -1;
+    unsigned int clear_mask = 0x00000011;
+    for (int pos = 1; pos < pattern_length; pos++, candidate++) {
+        unsigned char com_pos = (pattern[pos / 4] >> (2 * (pos % 4))) |= clear_mask;
```
Comment (xye16) [inline kmp_cpu.cpp:8]: Why "|=" instead of "|"?

File: kmp_cpu.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            text_com[i] |= (A_mask << (2 * j));
+        }
+        else if (text_tmp == 'T') {
+            text_com[i] |= (B_mask << (2 * j));
+        }
+        else if (text_tmp == 'C') {
+            text_com[i] |= (C_mask << (2 * j));
+        }
+        else if (text_tmp == 'G') {
+            text_com[i] |= (D_mask << (2 * j));
```
Comment (xye16) [inline kmp_cpu.cpp:43]: Typo: D_mask -> G_mask.

File: kmp_cpu.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    unsigned int G_mask = 0x00000011;
+    for (int idx = 0; idx < text_length; idx++) {
+        int i = idx / 4;
+        int j = idx % 4;
+        char text_tmp = text[4 * i + j];
+        if (text_tmp == 'A') {
+            text_com[i] |= (A_mask << (2 * j));
+        }
+        else if (text_tmp == 'T') {
+            text_com[i] |= (B_mask << (2 * j));
```
Comment (xye16) [inline kmp_cpu.cpp:37]: Typo: B_mask -> T_mask.

**Review by xye16 — COMMENTED** [review]
Not done yet, but here are some of the problems.

File: .DS_Store
[the remark this answers is not in the document]
Comment (Brack-Wang) [reply .DS_Store:1]: Deleted

File: brute_force.hpp
[the remark this answers is not in the document]
Comment (Brack-Wang) [reply brute_force.hpp:1]: OK don't know how to merge before. I will delete this version and push again later

File: kmp_cpu.cpp (hunk 1 above)
Comment (Brack-Wang) [reply kmp_cpu.cpp:8]: changed

File: kmp_cpu.cpp (hunk 2 above)
Comment (Brack-Wang) [reply kmp_cpu.cpp:43]: changed

File: kmp_cpu.cpp (hunk 3 above)
Comment (Brack-Wang) [reply kmp_cpu.cpp:37]: changed

**Comment by Brack-Wang** [comment]
Changed version. Compress

File: kmp_cpu.cpp [hunk 4]
```diff
@@ -3,30 +3,66 @@
...
+        else if (text_tmp == 'G') {
+            text_com[i] |= (G_mask << (2 * j));
+        }
+    }
+}
+
 int KMP_search(
-    const char *text, int text_length, const char *pattern, int pattern_length,
-    int *output, int max_output_cnt, int *fail
+    const char *text, int16 text_length, const char *pattern, int16 pattern_length,
```
Comment (xye16) [inline kmp_cpu.cpp:49]: text_length should be an int. It can go over 2^15.

Comment (xye16) [inline kmp_cpu.cpp:49]: The type of 16 bit int is called `int16_t`.

File: kmp_cpu.cpp [hunk 5]
```diff
@@ -3,30 +3,66 @@
 static void get_fail(const char *pattern, int pattern_length, int *fail) {
```
Comment (xye16) [inline kmp_cpu.cpp:3]: Change the type of pattern_length to int16_t, and fail to int16_t* for consistency.

File: kmp_cpu.cpp [hunk 6]
```diff
@@ -3,30 +3,66 @@
...
     int candidate = 0;
     fail[0] = -1;
+    unsigned char clear_mask = 0x00000011;
```
Comment (xye16) [inline kmp_cpu.cpp:6]: Should be `0b` instead of `0x`.

File: kmp_cpu.cpp [hunk 7]
```diff
@@ -3,30 +3,66 @@
...
     }
     fail[pattern_length] = candidate;
 }
 
+void char_compress(const char *text, int text_length, char& text_com) {
+    // A: 00; T: 01; C:10; G: 11 
+    unsigned char A_mask = 0x00000000;
+    unsigned char T_mask = 0x00000001;
+    unsigned char C_mask = 0x00000010;
+    unsigned char G_mask = 0x00000011;
```
Comment (xye16) [inline kmp_cpu.cpp:28]: All these masks should also use `0b` prefix instead of `0x`.

File: kmp_cpu.cpp [hunk 8]
```diff
@@ -3,30 +3,66 @@
...
+
+    get_fail(pattern_com, pattern_length, fail);
 
     int i = 0, j = 0;
     int output_cnt = 0;
+    unsigned char clear_mask = 0x00000011
     while (i < text_length) {
-        if (text[i] == pattern[j]) {
+        unsigned char com_text = (text_com[i / 4] >> (2 * (i % 4))) | clear_mask
+        unsigned char com_pat = (pattern_com[i / 4] >> (2 * (i % 4))) | clear_mask
```
Comment (xye16) [inline kmp_cpu.cpp:64]: `& clear_mask`

File: kmp_cpu.cpp
Comment (xye16) [inline kmp_cpu.cpp:1]: After you are done with all these modifications. Add this file into rai_build.yml, and run `rai -p .`. Make sure your code gets correct results on some sample test cases.

**Review by xye16 — CHANGES_REQUESTED** [review]

File: kmp_cpu.cpp (hunk 4 above)
Comment (Brack-Wang) [reply kmp_cpu.cpp:49]: changed

Comment (Brack-Wang) [reply kmp_cpu.cpp:49]: changed

File: kmp_cpu.cpp (hunk 5 above)
Comment (Brack-Wang) [reply kmp_cpu.cpp:3]: changed

File: kmp_cpu.cpp (hunk 6 above)
Comment (Brack-Wang) [reply kmp_cpu.cpp:6]: changed

File: kmp_cpu.cpp (hunk 7 above)
Comment (Brack-Wang) [reply kmp_cpu.cpp:28]: changed

File: kmp_cpu.cpp (hunk 8 above)
Comment (Brack-Wang) [reply kmp_cpu.cpp:64]: change | to &

**Review by xye16 — APPROVED** [review]

**Comment by Brack-Wang** [comment]
char compress function

**Pull request merged**